src/thor/timedep_forward.cc

```diff
@@ -0,0 +1,285 @@
+                   const uint32_t pred_idx, const bool from_transition,
+                   const uint32_t localtime,
+                   const odin::Location& destination,
+                   std::pair<int32_t, float>& best_path) {
+  // Get the tile and the node info. Skip if tile is null (can happen
+  // with regional data sets) or if no access at the node.
+  const GraphTile* tile = graphreader.GetGraphTile(node);
+  if (tile == nullptr) {
+    return;
+  }
+  const NodeInfo* nodeinfo = tile->node(node);
+  if (!costing_->Allowed(nodeinfo)) {
+    return;
+  }
+
+  // Expand from end node.
+  uint32_t shortcuts = 0;
+  uint32_t max_shortcut_length = static_cast<uint32_t>(pred.distance() * 0.5f);
+  GraphId edgeid(node.tileid(), node.level(), nodeinfo->edge_index());
+  EdgeStatusInfo* es = edgestatus_.GetPtr(edgeid, tile);
+  const DirectedEdge* directededge = tile->directededge(nodeinfo->edge_index());
+  for (uint32_t i = 0; i < nodeinfo->edge_count(); ++i, ++directededge, ++edgeid, ++es) {
+    // Handle transition edges - expand from the end node of the transition
+    // (unless this is called from a transition).
+    if (directededge->trans_up()) {
+      if (!from_transition) {
+        hierarchy_limits_[node.level()].up_transition_count++;
+        ExpandForward(graphreader, directededge->endnode(), pred, pred_idx,
+                      true, localtime, destination, best_path);
+      }
+      continue;
+    } else if (directededge->trans_down()) {
+      if (!from_transition &&
+          !hierarchy_limits_[directededge->endnode().level()].StopExpanding(pred.distance())) {
+        ExpandForward(graphreader, directededge->endnode(), pred, pred_idx,
+                      true, localtime, destination, best_path);
+      }
+      continue;
+    }
+
+    // Skip this edge if permanently labeled (best path already found to this
+    // directed edge), if edge is superseded by a shortcut edge that was taken,
+    // or if no access is allowed to this edge (based on costing method), or if
+    // a complex restriction exists.
+    if (es->set() == EdgeSet::kPermanent ||
+        (shortcuts & directededge->superseded()) ||
+       !costing_->Allowed(directededge, pred, tile, edgeid, localtime) ||
+        costing_->Restricted(directededge, pred, edgelabels_, tile,
+                                     edgeid, true, localtime)) {
+      continue;
+    }
+
+    // Skip shortcut edges when near the destination. Always skip within
+    // 10km but also reject long shortcut edges outside this distance.
+    // TODO - configure this distance based on density?
+    if (directededge->is_shortcut() && (pred.distance() < 10000.0f ||
+        directededge->length() > max_shortcut_length)) {
+      continue;
+    }
+
+    // Update the_shortcuts mask
+    shortcuts |= directededge->shortcut();
+
+    // Compute the cost to the end of this edge
+    Cost newcost = pred.cost() + costing_->EdgeCost(directededge) +
```

Comment:
The intention of this PR is just to enable time-dependent access restrictions, right? Eventually, I assume `localtime` would need to be passed to `EdgeCost` ?

Reply:
yes, but will likely have to pass in either the GraphTile and edge Id or other information to access any predictive speed / duration data (and real-time in the future). So waiting until we have more info until that interface gets defined.